enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,319 @@
+
+### Goals
+
+- Define an initial set of hook points for the deletion phase.
+- Define an initial set and form of related annotations.
+- Define basic expectations for a controller or process that responds to a
+lifecycle hook.
+
+### Non-Goals
+
+- Create an exhaustive list of hooks; we can add more over time.
+- Create new machine phases.
+- Create a mechanism to signal what lifecycle point a machine is at currently.
+- Dictate implementation of controllers that respond to the hooks.
+- Implement ordering in the machine controller.
+- Require anyone to use these hooks for normal machine operations, these are
+strictly optional and for custom integrations only.
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+### User Stories
+
+#### Story 1
+(pre-terminate) As an operator, I would like to have the ability to perform
+different actions between the time a Machine is marked deleted in the API and
+the time the Machine is deleted from the infrastructure provider.
+
+For example, when replacing a Control Plane Machine, ensure a new Control
+Plane Machine has been successfully created and joined to the cluster before
+removing the instance of the deleted Machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.  This will also prevent
+prolonged periods of having one fewer Control Plane host in the event the
+replacement instance does not come up in a timely manner.
+
+#### Story 2
+(pre-drain) As an operator, I want the ability to utilize my own draining
+controller instead of the logic built into the machine controller.  This will
+allow me better flexibility and control over the lifecycle of workloads on each
+node.
+
+### Implementation Details/Notes/Constraints
+
+For each defined lifecycle point, one or more hooks may be applied as an annotation
+to the machine object. These annotations will pause reconciliation of a Machine object
+until all hooks are resolved for that lifecycle point.
+The hooks should be managed by a Hook Implementing Controller or other external
+application, or manually created and removed by an administrator.
+
+#### Lifecycle Points
+
+##### pre-drain
+
+`pre-drain.delete.hook.machine.cluster.x-k8s.io`
+
+Hooks defined at this point will prevent the machine controller from draining a node
+after the machine object has been marked for deletion until the hooks are removed.
+
+##### pre-terminate
+
+`pre-terminate.delete.hook.machine.cluster.x-k8s.io`
+
+Hooks defined at this point will prevent the machine-controller from
+removing/terminating the instance in the cloud provider until the hooks are
+removed.
```

Comment:
how will we mesh this with quorum-guard and or other PDB. basically if PDB is not met drain will not happen. So the expectation would be after hook is removed then the node will drain. We need to have clear signal.

...hook removed but PBD not met.

Reply:
I think we can represent this in a condition in the Machine status if we feel a need. Feedback from MAPI today isn't great, this is known and we are working on this.

I think today if a drain is blocked because of a PDB, we log a message but don't surface this onto a machine.

I could see us adding two conditions off the back of this proposal maybe, a terminable condition that reflects if pre-terminate hooks are present, and a drainable hook that reflects whether pre-drain hooks are present. For the drain situation itself, we could add a drained condition that could reflect any errors or blocks in the draining process when a Machine has a deletion timestamp